exp/runtime/internal/controllers/extension_controller.go

```diff
@@ -21,74 +21,169 @@ import (
+	err := r.Client.Get(ctx, req.NamespacedName, extension)
 	if err != nil {
 		if apierrors.IsNotFound(err) {
-			// Object not found, return.  Created objects are automatically garbage collected.
-			// For additional cleanup logic use finalizers.
+			// Object not found, return.
 			return ctrl.Result{}, nil
 		}
 
 		// Error reading the object - requeue the request.
 		return ctrl.Result{}, err
 	}
 
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(extension, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// FIXME: could dedupe these patch calls. Ignoring for simplicity at this stage.
+	defer func() {
+		// Always attempt to patch the object and status after each reconciliation.
+		// Patch ObservedGeneration only if the reconciliation completed successfully
+		patchOpts := []patch.Option{}
+		if reterr == nil {
+			patchOpts = append(patchOpts, patch.WithStatusObservedGeneration{},
+				patch.WithOwnedConditions{Conditions: []clusterv1.ConditionType{
+					runtimev1.RuntimeExtensionDiscovered,
+				}})
+		}
+		if err := patchExtension(ctx, patchHelper, extension, patchOpts...); err != nil {
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
 	// Handle deletion reconciliation loop.
-	if !ext.ObjectMeta.DeletionTimestamp.IsZero() {
-		return r.reconcileDelete(ctx, ext)
+	if !extension.ObjectMeta.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(ctx, extension)
 	}
 
 	// Handle normal reconciliation loop.
-	return r.reconcile(ctx, ext)
+	return r.reconcile(ctx, extension)
 }
 
-func (r *ExtensionReconciler) reconcileDelete(ctx context.Context, ext *expv1.Extension) (ctrl.Result, error) {
-	r.Registry.RemoveRuntimeExtension(ext)
+func patchExtension(ctx context.Context, patchHelper *patch.Helper, ext *runtimev1.Extension, options ...patch.Option) error {
+	// Can add errors and waits here to the Extension object.
+	return patchHelper.Patch(ctx, ext, options...)
+}
 
+func (r *Reconciler) reconcileDelete(_ context.Context, extension *runtimev1.Extension) (ctrl.Result, error) {
+	// FIXME: This isn't implemented on the client/registry.
+	if err := r.RuntimeClient.Extension(extension).Unregister(); err != nil {
+		return ctrl.Result{}, err
+	}
 	return ctrl.Result{}, nil
 }
 
-func (r *ExtensionReconciler) reconcile(ctx context.Context, ext *expv1.Extension) (ctrl.Result, error) {
-	// FIXME:
-	// * should call Discover on RuntimeExtension and then update Status
-	// Note: Has to work with ext.Spec.ClientConfig without underlying registry
-	// Q: Is it enough to do the Discovery only once initially or on each reconcile
+func (r *Reconciler) reconcile(ctx context.Context, extension *runtimev1.Extension) (ctrl.Result, error) {
+	if err := r.discoverRuntimeExtensions(ctx, extension); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
 
-	r.Registry.RegisterRuntimeExtension(ext)
+// FIXME: should this method use an api reader to ensure we don't hit the cache?
```

Comment:
Hm yeah let's use the apiReader we only actually do this call once so the overhead is negligible. The impact of missing a few Extensions is to problematic.